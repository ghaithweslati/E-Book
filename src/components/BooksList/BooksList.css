.listLivre
{
  padding: 0% 5%;
  margin-top: 4%;
}


.listLivre header>h1
{
  padding: 0;
  color: #484848;
  font-weight: normal;
  font-size: 28px;
  float: left;
  margin: 0;
}

.listLivre header>h1+h1
{
  margin-right: 7%;
  float: right;
  border: #b1b0b0 solid 0.5px;
  color: #b1b0b0;
  border-radius: 50%;
  padding: 5px;
  width: 30px;
  text-align: center;
}

.listLivre header>h1+h1:hover
{
  background-color:#77b748;
  color: white;
}

.listLivre hr
{
width: 96%;
text-align:left;
margin-left:0;
border:  solid 0.5px #dadada;
margin-bottom: 2.5%;

}

.listLivre section>*
{
  display: inline-table;
}

.listLivre header nav>input
{
  
    width: 300%;
    height: 10%;
    font-size: 15px;
    margin-left: 10%;
    text-align: center;
    border: solid 1px rgb(190, 189, 189);
    padding:10px;
    text-align: left;
    margin-left: 30px;
    margin-top: -50px;
}

.listLivre header>*
{
display: inline-table;
}
.listLivre header
{
  padding: 0;
  margin: 0;
  margin-bottom: 3%;
}