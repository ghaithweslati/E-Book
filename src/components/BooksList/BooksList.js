import Book from "../Book/Book";
import "./BooksList.css";
import BookForm from "../BookForm/BookForm";
import React, { useState, useCallback, useEffect } from "react";
import { fetchBooks, searchBooks } from "../../services/books.service";

function BooksList(props) {
  const [books, setBooks] = useState([]);
  const [searchValue, setSearchValue] = useState("");
  const [add, setAdd] = useState(true);

  var current = { id: 0, type: "Admin" };
  if (localStorage.getItem("current"))
    current = JSON.parse(localStorage.getItem("current"));

  const addBook = (image, categorie, titre, nbExemp, auteur) => {
    setBooks((previousBooks) => [
      {
        isbn: books.length + 1,
        image,
        categorie,
        titre,
        etat: "Disponible",
        nbExemp,
        auteur,
      },
      ...previousBooks,
    ]);
    fetchBooks().unshift({
      isbn: fetchBooks().length + 1,
      image,
      categorie,
      titre,
      etat: "Disponible",
      nbExemp,
      auteur,
    });
    localStorage.setItem("books", JSON.stringify(fetchBooks()));
  };

  const memoizedCallback = useCallback(addBook, []);

  const updateBook = (isbn, image, categorie, titre, etat, nbExemp, auteur) => {
    const newBooks = books.map((book) =>
      book.isbn === isbn
        ? { isbn, image, categorie, titre, etat, nbExemp, auteur }
        : book
    );
    setBooks(newBooks);
    localStorage.setItem("books", JSON.stringify(newBooks));
  };

  useEffect(() => {
    const fetchData = async () => {
      if (!searchValue) {
        let result = await fetchBooks();
        setBooks(result);
      } else {
        const result = await searchBooks(searchValue);
        setBooks(result);
      }
    };
    fetchData();
  }, [searchValue]);

  return (
    <div className="listLivre">
      <header>
        <h1>List of Books</h1>
        {current.type === "Admin" && <h1 onClick={(e) => setAdd(!add)}>+</h1>}
        {
          <nav>
            {add ? (
              <input
                type="text"
                placeholder="Search By Title, Author..."
                name="search"
                value={searchValue}
                onChange={(e) => setSearchValue(e.target.value)}
              />
            ) : (
              <BookForm addBook={memoizedCallback} maxnbExempValue={100} />
            )}
          </nav>
        }
      </header>
      <hr />
      <section>
        {books.map((step, index) => (
          <Book
            key={index}
            isbn={step.isbn}
            image={step.image}
            categorie={step.categorie}
            titre={step.titre}
            etat={step.etat}
            nbExemp={step.nbExemp}
            auteur={step.auteur}
            updateBook={updateBook}
            getBook={props.getBook}
            borrowed={props.borrowed}
          >
            {step.description}
          </Book>
        ))}
      </section>
    </div>
  );
}

export default BooksList;
