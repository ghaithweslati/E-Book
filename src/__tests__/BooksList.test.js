import React from "react";
import { render, act } from "@testing-library/react";
import BooksList from "../components/BooksList/BooksList";
// import { mockBooks } from "../mock/mockBooks";

describe("testing books list", () => {
  const mockBorrowed = 1;

  test("should render an array of books list", async () => {
    const promise = Promise.resolve();
    const mockGetBook = jest.fn(() => promise);
    const BooksListComponent = render(
      <BooksList borrowed={mockBorrowed} getBook={mockGetBook} />
    );
    await act(() => promise);
    expect(BooksListComponent).toMatchSnapshot();
  });
});
