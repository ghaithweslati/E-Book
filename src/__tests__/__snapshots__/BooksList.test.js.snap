// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing books list should render an array of books list 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="listLivre"
      >
        <header>
          <h1>
            List of Books
          </h1>
          <nav>
            <input
              name="search"
              placeholder="Search By Title, Author..."
              type="text"
              value=""
            />
          </nav>
        </header>
        <hr />
        <section>
          <div
            class="book"
          >
            <img
              alt="Help Me FindMy Stomach"
              src="5.PNG"
            />
            <span>
              Art & Photography
            </span>
            <hr />
            <label
              id="Disponible"
            >
              Disponible
            </label>
            <b
              data-testid="titre-book"
            >
              Help Me FindMy Stomach
            </b>
            <b />
            <span>
              By 
              Angela Gunning
            </span>
            <h5>
              2
               Copies
            </h5>
            <p>
              Emprunter
            </p>
          </div>
          <div
            class="book"
          >
            <img
              alt="Drive Safely,No Bumping"
              src="2.PNG"
            />
            <span>
              Art & Photography
            </span>
            <hr />
            <label
              id="Disponible"
            >
              Disponible
            </label>
            <b
              data-testid="titre-book"
            >
              Drive Safely,No Bumping
            </b>
            <b />
            <span>
              By 
              Angela Gunning
            </span>
            <h5>
              2
               Copies
            </h5>
            <p>
              Emprunter
            </p>
          </div>
          <div
            class="book"
          >
            <img
              alt="Let the Good Times Roll Up"
              src="3.PNG"
            />
            <span>
              Art & Photography
            </span>
            <hr />
            <label
              id="Disponible"
            >
              Disponible
            </label>
            <b
              data-testid="titre-book"
            >
              Let the Good Times Roll Up
            </b>
            <b />
            <span>
              By 
              Angela Gunning
            </span>
            <h5>
              2
               Copies
            </h5>
            <p>
              Emprunter
            </p>
          </div>
          <div
            class="book"
          >
            <img
              alt="Out State Fair Is A Great State Fair"
              src="4.PNG"
            />
            <span>
              Art & Photography
            </span>
            <hr />
            <label
              id="Disponible"
            >
              Disponible
            </label>
            <b
              data-testid="titre-book"
            >
              Out State Fair Is A Great State Fair
            </b>
            <b />
            <span>
              By 
              Angela Gunning
            </span>
            <h5>
              2
               Copies
            </h5>
            <p>
              Emprunter
            </p>
          </div>
          <div
            class="book"
          >
            <img
              alt="Put The Petal To The Metal"
              src="5.PNG"
            />
            <span>
              Art & Photography
            </span>
            <hr />
            <label
              id="Disponible"
            >
              Disponible
            </label>
            <b
              data-testid="titre-book"
            >
              Put The Petal To The Metal
            </b>
            <b />
            <span>
              By 
              Angela Gunning
            </span>
            <h5>
              2
               Copies
            </h5>
            <p>
              Emprunter
            </p>
          </div>
          <div
            class="book"
          >
            <img
              alt="Help Me Find My Stomach"
              src="3.PNG"
            />
            <span>
              Art & Photography
            </span>
            <hr />
            <label
              id="Disponible"
            >
              Disponible
            </label>
            <b
              data-testid="titre-book"
            >
              Help Me Find My Stomach
            </b>
            <b />
            <span>
              By 
              Angela Gunning
            </span>
            <h5>
              2
               Copies
            </h5>
            <p>
              Emprunter
            </p>
          </div>
          <div
            class="book"
          >
            <img
              alt="Help Me Find My Stomach"
              src="4.PNG"
            />
            <span>
              Art & Photography
            </span>
            <hr />
            <label
              id="Disponible"
            >
              Disponible
            </label>
            <b
              data-testid="titre-book"
            >
              Help Me Find My Stomach
            </b>
            <b />
            <span>
              By 
              Angela Gunning
            </span>
            <h5>
              2
               Copies
            </h5>
            <p>
              Emprunter
            </p>
          </div>
          <div
            class="book"
          >
            <img
              alt="Help Me Find My Stomach"
              src="5.PNG"
            />
            <span>
              Art & Photography
            </span>
            <hr />
            <label
              id="Disponible"
            >
              Disponible
            </label>
            <b
              data-testid="titre-book"
            >
              Help Me Find My Stomach
            </b>
            <b />
            <span>
              By 
              Angela Gunning
            </span>
            <h5>
              2
               Copies
            </h5>
            <p>
              Emprunter
            </p>
          </div>
          <div
            class="book"
          >
            <img
              alt="Help Me Find My Stomach"
              src="2.PNG"
            />
            <span>
              Art & Photography
            </span>
            <hr />
            <label
              id="Disponible"
            >
              Disponible
            </label>
            <b
              data-testid="titre-book"
            >
              Help Me Find My Stomach
            </b>
            <b />
            <span>
              By 
              Angela Gunning
            </span>
            <h5>
              2
               Copies
            </h5>
            <p>
              Emprunter
            </p>
          </div>
          <div
            class="book"
          >
            <img
              alt="Help Me Find My Stomach"
              src="3.PNG"
            />
            <span>
              Art & Photography
            </span>
            <hr />
            <label
              id="Disponible"
            >
              Disponible
            </label>
            <b
              data-testid="titre-book"
            >
              Help Me Find My Stomach
            </b>
            <b />
            <span>
              By 
              Angela Gunning
            </span>
            <h5>
              2
               Copies
            </h5>
            <p>
              Emprunter
            </p>
          </div>
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="listLivre"
    >
      <header>
        <h1>
          List of Books
        </h1>
        <nav>
          <input
            name="search"
            placeholder="Search By Title, Author..."
            type="text"
            value=""
          />
        </nav>
      </header>
      <hr />
      <section>
        <div
          class="book"
        >
          <img
            alt="Help Me FindMy Stomach"
            src="5.PNG"
          />
          <span>
            Art & Photography
          </span>
          <hr />
          <label
            id="Disponible"
          >
            Disponible
          </label>
          <b
            data-testid="titre-book"
          >
            Help Me FindMy Stomach
          </b>
          <b />
          <span>
            By 
            Angela Gunning
          </span>
          <h5>
            2
             Copies
          </h5>
          <p>
            Emprunter
          </p>
        </div>
        <div
          class="book"
        >
          <img
            alt="Drive Safely,No Bumping"
            src="2.PNG"
          />
          <span>
            Art & Photography
          </span>
          <hr />
          <label
            id="Disponible"
          >
            Disponible
          </label>
          <b
            data-testid="titre-book"
          >
            Drive Safely,No Bumping
          </b>
          <b />
          <span>
            By 
            Angela Gunning
          </span>
          <h5>
            2
             Copies
          </h5>
          <p>
            Emprunter
          </p>
        </div>
        <div
          class="book"
        >
          <img
            alt="Let the Good Times Roll Up"
            src="3.PNG"
          />
          <span>
            Art & Photography
          </span>
          <hr />
          <label
            id="Disponible"
          >
            Disponible
          </label>
          <b
            data-testid="titre-book"
          >
            Let the Good Times Roll Up
          </b>
          <b />
          <span>
            By 
            Angela Gunning
          </span>
          <h5>
            2
             Copies
          </h5>
          <p>
            Emprunter
          </p>
        </div>
        <div
          class="book"
        >
          <img
            alt="Out State Fair Is A Great State Fair"
            src="4.PNG"
          />
          <span>
            Art & Photography
          </span>
          <hr />
          <label
            id="Disponible"
          >
            Disponible
          </label>
          <b
            data-testid="titre-book"
          >
            Out State Fair Is A Great State Fair
          </b>
          <b />
          <span>
            By 
            Angela Gunning
          </span>
          <h5>
            2
             Copies
          </h5>
          <p>
            Emprunter
          </p>
        </div>
        <div
          class="book"
        >
          <img
            alt="Put The Petal To The Metal"
            src="5.PNG"
          />
          <span>
            Art & Photography
          </span>
          <hr />
          <label
            id="Disponible"
          >
            Disponible
          </label>
          <b
            data-testid="titre-book"
          >
            Put The Petal To The Metal
          </b>
          <b />
          <span>
            By 
            Angela Gunning
          </span>
          <h5>
            2
             Copies
          </h5>
          <p>
            Emprunter
          </p>
        </div>
        <div
          class="book"
        >
          <img
            alt="Help Me Find My Stomach"
            src="3.PNG"
          />
          <span>
            Art & Photography
          </span>
          <hr />
          <label
            id="Disponible"
          >
            Disponible
          </label>
          <b
            data-testid="titre-book"
          >
            Help Me Find My Stomach
          </b>
          <b />
          <span>
            By 
            Angela Gunning
          </span>
          <h5>
            2
             Copies
          </h5>
          <p>
            Emprunter
          </p>
        </div>
        <div
          class="book"
        >
          <img
            alt="Help Me Find My Stomach"
            src="4.PNG"
          />
          <span>
            Art & Photography
          </span>
          <hr />
          <label
            id="Disponible"
          >
            Disponible
          </label>
          <b
            data-testid="titre-book"
          >
            Help Me Find My Stomach
          </b>
          <b />
          <span>
            By 
            Angela Gunning
          </span>
          <h5>
            2
             Copies
          </h5>
          <p>
            Emprunter
          </p>
        </div>
        <div
          class="book"
        >
          <img
            alt="Help Me Find My Stomach"
            src="5.PNG"
          />
          <span>
            Art & Photography
          </span>
          <hr />
          <label
            id="Disponible"
          >
            Disponible
          </label>
          <b
            data-testid="titre-book"
          >
            Help Me Find My Stomach
          </b>
          <b />
          <span>
            By 
            Angela Gunning
          </span>
          <h5>
            2
             Copies
          </h5>
          <p>
            Emprunter
          </p>
        </div>
        <div
          class="book"
        >
          <img
            alt="Help Me Find My Stomach"
            src="2.PNG"
          />
          <span>
            Art & Photography
          </span>
          <hr />
          <label
            id="Disponible"
          >
            Disponible
          </label>
          <b
            data-testid="titre-book"
          >
            Help Me Find My Stomach
          </b>
          <b />
          <span>
            By 
            Angela Gunning
          </span>
          <h5>
            2
             Copies
          </h5>
          <p>
            Emprunter
          </p>
        </div>
        <div
          class="book"
        >
          <img
            alt="Help Me Find My Stomach"
            src="3.PNG"
          />
          <span>
            Art & Photography
          </span>
          <hr />
          <label
            id="Disponible"
          >
            Disponible
          </label>
          <b
            data-testid="titre-book"
          >
            Help Me Find My Stomach
          </b>
          <b />
          <span>
            By 
            Angela Gunning
          </span>
          <h5>
            2
             Copies
          </h5>
          <p>
            Emprunter
          </p>
        </div>
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
